File: coding_skill_py/03_singleton_pattern.py
"""
Python Singleton Pattern - 메타클래스와 데코레이터
C 버전: static 변수 + 초기화 플래그
Python 버전: 메타클래스, __new__, 데코레이터
"""

from functools import wraps
from threading import Lock

# 1. 메타클래스를 이용한 Singleton
class SingletonMeta(type):
    _instances = {}
    _lock = Lock()
    
    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            with cls._lock:
                if cls not in cls._instances:
                    cls._instances[cls] = super().__call__(*args, **kwargs)
        return cls._instances[cls]

class Logger(metaclass=SingletonMeta):
    def __init__(self):
        print("[Logger] 생성")
    
    def log(self, message: str):
        print(f"[LOG] {message}")

# 2. __new__를 이용한 Singleton
class Config:
    _instance = None
    
    def __new__(cls):
        if cls._instance is None:
            print("[Config] 생성")
            cls._instance = super().__new__(cls)
            cls._instance.version = "1.0.0"
        return cls._instance
    
    def get_version(self) -> str:
        return self.version

# 3. 데코레이터를 이용한 Singleton
def singleton(cls):
    instances = {}
    lock = Lock()
    
    @wraps(cls)
    def get_instance(*args, **kwargs):
        if cls not in instances:
            with lock:
                if cls not in instances:
                    instances[cls] = cls(*args, **kwargs)
        return instances[cls]
    
    return get_instance

@singleton
class Database:
    def __init__(self):
        print("[Database] 생성")
    
    def query(self, sql: str):
        print(f"[DB] Query: {sql}")

# 4. 모듈 레벨 Singleton (가장 Pythonic)
# Python에서 모듈은 기본적으로 싱글톤!
class _SystemMonitor:
    def __init__(self):
        print("[SystemMonitor] 생성")
        self.cpu_usage = 0.0
    
    def update(self):
        self.cpu_usage += 1.0
        print(f"[Monitor] CPU: {self.cpu_usage}%")

# 모듈 레벨에서 인스턴스 생성
system_monitor = _SystemMonitor()

def main():
    print("\n=== Python Singleton Pattern ===")
    
    # 메타클래스 Singleton
    print("\n1. 메타클래스 Singleton:")
    logger1 = Logger()
    logger2 = Logger()
    print(f"Same instance? {logger1 is logger2}")
    logger1.log("Message 1")
    
    # __new__ Singleton
    print("\n2. __new__ Singleton:")
    config1 = Config()
    config2 = Config()
    print(f"Same instance? {config1 is config2}")
    print(f"Version: {config1.get_version()}")
    
    # 데코레이터 Singleton
    print("\n3. 데코레이터 Singleton:")
    db1 = Database()
    db2 = Database()
    print(f"Same instance? {db1 is db2}")
    db1.query("SELECT * FROM users")
    
    # 모듈 레벨 Singleton
    print("\n4. 모듈 레벨 Singleton:")
    system_monitor.update()
    system_monitor.update()
    
    print("\n모든 방법으로 싱글톤 구현 가능!")

if __name__ == "__main__":
    main()

File: coding_skill_cpp/03_singleton_pattern.cpp
/*
 * C++ Singleton Pattern - Meyers Singleton (Thread-safe)
 * C 버전: static 변수 + 초기화 플래그
 * C++ 버전: 지역 static (C++11 이후 thread-safe 보장)
 */

#include <iostream>
#include <memory>
#include <mutex>
#include <string>

using namespace std;

namespace patterns {

// 1. Meyers Singleton (가장 권장)
class Logger {
    Logger() { cout << "[Logger] 생성" << endl; }
    
public:
    // 복사/이동 방지
    Logger(const Logger&) = delete;
    Logger& operator=(const Logger&) = delete;
    
    static Logger& getInstance() {
        static Logger instance;  // C++11: thread-safe!
        return instance;
    }
    
    void log(const string& msg) {
        cout << "[LOG] " << msg << endl;
    }
};

// 2. CRTP를 이용한 Singleton Base
template<typename T>
class Singleton {
protected:
    Singleton() = default;
    
public:
    Singleton(const Singleton&) = delete;
    Singleton& operator=(const Singleton&) = delete;
    
    static T& getInstance() {
        static T instance;
        return instance;
    }
};

class Config : public Singleton<Config> {
    friend class Singleton<Config>;
    string version = "1.0.0";
    
    Config() { cout << "[Config] 생성" << endl; }
    
public:
    string getVersion() const { return version; }
    void setVersion(const string& v) { version = v; }
};

// 3. Thread-safe Singleton (명시적 mutex)
class Database {
    static unique_ptr<Database> instance;
    static mutex mtx;
    
    Database() { cout << "[Database] 생성" << endl; }
    
public:
    static Database& getInstance() {
        if (!instance) {
            lock_guard<mutex> lock(mtx);
            if (!instance) {
                instance = unique_ptr<Database>(new Database());
            }
        }
        return *instance;
    }
    
    void query(const string& sql) {
        cout << "[DB] Query: " << sql << endl;
    }
};

unique_ptr<Database> Database::instance = nullptr;
mutex Database::mtx;

} // namespace patterns

int main() {
    using namespace patterns;
    
    cout << "\n=== C++ Singleton Pattern ===" << endl;
    
    // Meyers Singleton
    cout << "\n1. Meyers Singleton:" << endl;
    Logger::getInstance().log("Message 1");
    Logger::getInstance().log("Message 2");
    
    // CRTP Singleton
    cout << "\n2. CRTP Singleton:" << endl;
    cout << "Version: " << Config::getInstance().getVersion() << endl;
    
    // Thread-safe Singleton
    cout << "\n3. Thread-safe Singleton:" << endl;
    Database::getInstance().query("SELECT * FROM users");
    
    cout << "\n모두 동일한 인스턴스 사용!" << endl;
    
    return 0;
}
